packages/contracts/contracts/updaters/OZUUPSUpdater.sol

```diff
@@ -0,0 +1,133 @@
+    /**
+     * @notice An event that is emitted each time the implementation is changed. This event is part
+     *         of the EIP-1967 specification.
+     *
+     * @param implementation The address of the implementation contract
+     */
+    event Upgraded(address indexed implementation);
+
+    function proxiableUUID() external view virtual returns (bytes32) {
+        return 0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+    }
+
+    /**
+     * @notice Set the implementation contract address. The code at the given address will execute
+     *         when this contract is called.
+     *
+     * @param _implementation Address of the implementation contract.
+     */
+    function upgradeTo(address _implementation) external ifChugSplashAdmin {
+        _setImplementation(_implementation);
+    }
+
+    /**
+     * @notice Sets the implementation address.
+     *
+     * @param _implementation New implementation address.
+     */
+    function _setImplementation(address _implementation) internal {
+        assembly {
+            sstore(IMPLEMENTATION_KEY, _implementation)
+        }
+        emit Upgraded(_implementation);
+    }
+
+    /**
+     * @notice Sets the chugsplash specific admin address.
+     *
+     * @param _newAdmin New admin address.
+     */
+    function _setChugSplashAdmin(address _newAdmin) internal {
+        assembly {
+            sstore(CHUGSPLASH_ADMIN_KEY, _newAdmin)
+        }
+    }
+
+    /**
+     * @notice Queries the chugsplash specific admin address.
+     *
+     * @return Owner address.
+     */
+    function _getChugSplashAdmin() internal view returns (address) {
+        address chugsplashAdmin;
+        assembly {
+            chugsplashAdmin := sload(CHUGSPLASH_ADMIN_KEY)
+        }
+        return chugsplashAdmin;
+    }
+
+    /**
+     * @notice A modifier that reverts if not called by the owner or by address(0) to allow
+     *         eth_call to interact with this proxy without needing to use low-level storage
+     *         inspection. We assume that nobody is able to trigger calls from address(0) during
+     *         normal EVM execution.
+     */
+    modifier ifChugSplashAdmin() {
+        if (msg.sender == _getChugSplashAdmin() || msg.sender == address(0)) {
+            _;
+        } else {
+            revert("unauthorized");
+        }
+    }
+
+    /**
+     * @notice Modifies some storage slot within the proxy contract. Gives us a lot of power to
+     *         perform upgrades in a more transparent way.
+     *
+     * @param _key   Storage key to modify.
+     * @param _value New value for the storage key.
+     */
+    function setStorage(bytes32 _key, bytes32 _value) external ifChugSplashAdmin {
+        assembly {
+            sstore(_key, _value)
+        }
+    }
+
+    /**
+     * @notice Sets up the proxy updater. In this case, there is no setup required.
+     */
+    function setup() public {
+        if (_getChugSplashAdmin() == address(0)) {
```

Comment:
nit: I think we can remove `if (_getChugSplashAdmin() == address(0))`. Will remove the need to do one unit test :)

Reply:
This contract actually can't be publicly accessible across multiple blocks because the `initiate` call in `OZUUPSAdapter.initiateExecution` always occurs immediately after the `upgradeTo` call. Even if it did occur across multiple blocks though, the if-statement doesn't add any extra security, because `_getChugSplashAdmin()` always returns `address(0)`